core/src/main/scala/algebra/ring/EuclideanRing.scala

```diff
@@ -0,0 +1,59 @@
+package algebra
+package ring
+
+import scala.annotation.tailrec
+import scala.{specialized => sp}
+
+/**
+ * EuclideanRing implements a Euclidean domain.
+ *
+ * The formal definition says that every euclidean domain A has (at
+ * least one) euclidean function f: A -> N (the natural numbers) where:
+ *
+ *   (for every x and non-zero y) x = yq + r, and r = 0 or f(r) < f(y).
+ *
+ * This generalizes the Euclidean division of integers, where f represents
+ * a measure of length (or absolute value), and the previous equation
+ * represents finding the quotient and remainder of x and y. So:
+ *
+ *   quot(x, y) = q
+ *   mod(x, y) = r
+ */
+trait EuclideanRing[@sp(Int, Long, Float, Double) A] extends Any with GCDRing[A] { self =>
```

Comment:
If we have specialized on these types can we also add the implementations?

Reply:
... Float and Double are taken charge of by Field (euclidean division is trivial).

For Byte/Short/Int/Long, the instances are not lawful as overflow is undefined behaviour territory (which was not the case for just `CommutativeRing`, the laws pass).

We could:

1. Provide no instances
2. Provide instances, but not test them
3. Port the `forAllSafe` machinery from Spire that enables tests with abort-on-overflow

What do you think?